dhis-2/dhis-services/dhis-service-administration/src/main/java/org/hisp/dhis/maintenance/jdbc/JdbcMaintenanceStore.java

```diff
@@ -304,6 +304,36 @@ public int deleteSoftDeletedTrackedEntityInstances()
         return result;
     }
 
+    /**
+     * This SQL needs to consider any other table that has a foreign key
+     * reference to the period table. Following the DHIS2 convention this should
+     * mean they have a column named {@code periodid}. When such tables are
+     * dropped they need to be removed from the union here. When such new tables
+     * are added they need to be added to the union here.
+     */
+    @Override
+    public void prunePeriods()
+    {
+        String sql = "with period_ids(periodid) as ("
+            + "    select distinct periodid from completedatasetregistration"
+            + "    union select distinct periodid from dataapproval"
+            + "    union select distinct periodid from dataapprovalaudit"
+            + "    union select distinct periodid from datainputperiod"
+            + "    union select distinct periodid from datavalue"
+            + "    union select distinct periodid from datavalueaudit"
+            + "    union select distinct periodid from eventchart_periods"
+            + "    union select distinct periodid from eventreport_periods"
+            + "    union select distinct periodid from eventvisualization_periods"
+            + "    union select distinct periodid from interpretation where periodid is not null"
+            + "    union select distinct periodid from lockexception where periodid is not null"
```

Comment:
Did you check if could move the null check to the main subselect, meaning `select periodid from period_ids where periodid is not null`?

Reply:
That works too.